src/functions.c

```diff
@@ -639,40 +773,58 @@ int functionsCreateWithFunctionCtx(sds function_name,sds engine_name, sds desc,
     }
     engine *engine = ei->engine;
 
-    functionInfo *fi = dictFetchValue(functions->functions, function_name);
-    if (fi && !replace) {
-        *err = sdsnew("Function already exists");
+    libraryInfo *li = dictFetchValue(lib_ctx->libraries, lib_name);
+    if (li && !replace) {
+        *err = sdsnew("Library already exists");
         return C_ERR;
     }
 
-    void *function = engine->create(engine->engine_ctx, code, err);
-    if (*err) {
-        return C_ERR;
+    if (li) {
+        libraryUnlink(lib_ctx, li);
     }
 
-    if (fi) {
-        /* free the already existing function as we are going to replace it */
-        engineFunctionFree(fi, functions);
+    libraryInfo *new_li = engineLibraryCreate(lib_name, ei, desc, code);
+    if (engine->create(engine->engine_ctx, new_li, code, err) != C_OK) {
+        goto error;
     }
 
-    engineFunctionCreate(function_name, function, ei, desc, code, functions);
+    if (dictSize(new_li->functions) == 0) {
+        *err = sdsnew("No libraries registered");
+        goto error;
+    }
+
+    if (libraryLink(lib_ctx, new_li, err) != C_OK) {
+        goto error;
+    }
+
+    if (li) {
+        engineLibraryFree(li);
+    }
 
     return C_OK;
+
+error:
+    engineLibraryFree(new_li);
+    if (li) {
+        int ret = libraryLink(lib_ctx, li, NULL);
```

Comment:
if we do add a flag to skip the collision check, here's another place to use it.

Reply:
Do you think its that critical? It should not happened a lot and it can actually help spot bugs (the return value is asserted).

Reply:
now it only happens in the non-happy path, but still there could be many (hundreds of) functions.
so libraryLink has two loops on that entire list (one for re-insertion, which we can't avoid), and the other one for collision check.
i think that adding an argument and a single line `if` is a very low price to pay to avoid an excessive loop which could maybe be big.

Reply:
@oranagra because there was only one place that was actually needed this verification, I moved it outside of the `libraryLink` function. Let me know what you think.

Reply:
LGTM